### ISSUE TITLE
Will the library support Svelte 5?

### ISSUE THREAD

**Issue by 0x5374656c6c6172** [ItalyPaleAle/svelte-spa-router#318]

**Comment by carbogninalberto** [comment]
We will make it happen at some point if @ItalyPaleAle approves it!

**Comment by ItalyPaleAle** [comment]
Yes support for Svelte 5 is definitely a goal. I'll look into it when the Svelte 5 project becomes a bit more stable

**Comment by ItalyPaleAle** [comment]
Yes support for Svelte 5 is definitely a goal. I'll look into it when the Svelte 5 project becomes a bit more stable

(And of course, PRs are welcome :) )

**Comment by carbogninalberto** [comment]
From this page @tgf9 you can also monitor the actual status: [https://svelte-5-preview.vercel.app/status](https://svelte-5-preview.vercel.app/status) 

We are already working on the migration to make it available as soon as v5 is ready 👍

**Comment by ItalyPaleAle** [comment]
Should have something ready in the next few days :)

I believe the router should still work with v5 for now, temporarily, thanks to the backwards-compatibility

**Comment by oneezy** [comment]
> Should have something ready in the next few days :)

any updates on this?

**Comment by carbogninalberto** [comment]
Svelte has been officially release, we are looking forward in the next few week to support it!

**Comment by carbogninalberto** [comment]
I just looked into it, if you upgrade right now to svelte 5 it will work fine, I didn't find any issue this thanks to the backward compatibility, but if you want to be sure wait a little bit until I prepare the PR :)

Quick note:
if you absolutely need to upgrade to access the new features like the snippets, for a while set this in your vite.config.js:

`compilerOptions: {
            compatibility: {
                componentApi: 4
            }
},
`

Stay tuned!

**Comment by nyhr** [comment]
Svelte 5 is not compatible with svelte-spa-router/wrap

'Type 'Component<$$ComponentProps, {}, "">' is not assignable to type 'ComponentType'.
  Type 'Component<$$ComponentProps, {}, "">' is not assignable to type 'new (options: ComponentConstructorOptions<Record<string, any>>) => SvelteComponent<Record<string, any>, any, any>'.
    Type 'Component<$$ComponentProps, {}, "">' provides no match for the signature 'new (options: ComponentConstructorOptions<Record<string, any>>): SvelteComponent<Record<string, any>, any, any>'

**Comment by ItalyPaleAle** [comment]
Hey folks, thanks for your patience. I've been incredibly busy with my day job and I'm currently on vacation. I'm hoping to get this done in the next week.

**Comment by SaadiSave** [comment]
@carbogninalberto Could you please explain how you got it working on Svelte 5? I am trying this, but it does not work:

### `BlogPost.svelte`

```svelte
<script lang="ts">
    import { onMount } from "svelte"
    import preprocessPost from "../lib/preprocess-post"

    let { path: string } = $props()

    let content = $state("")

    onMount(async () => {
        const post = await fetch(path)

        content = await post.text()

        content = preprocessPost(content)
    })
</script>

{@html content}
```

### App.svelte

```svelte
<script lang="ts">
  import BlogPost from "./components/BlogPost.svelte"
  import Router from "svelte-spa-router"
  import posts from "./lib/posts"

  const routes = {
      '/posts/:path': BlogPost,
  }
</script>

<ul>
    {#each posts as post}
        <li><a href="#/{post.path}">{post.title}</a></li>
    {/each}
</ul>

<Router routes={routes} />
```

I am trying to bind `path` and use it to dynamically load a blog post

**Comment by carbogninalberto** [comment]
on this line update from:

<li><a href="#/blog/{post.path}">{post.title}</a></li>

to

<li><a href="/#/blog/{post.path}">{post.title}</a></li>


or 

<li><a href="/blog/{post.path}" use:link>{post.title}</a></li>

**Comment by betim** [comment]
`export let params = {}` works as long as runes are not used. The moment `$state()` or any rune is present then it breaks.

A workaround:
```
// http://localhost:5173/#/admin/order/20
let params = window.location.href.split("/");
let id = params[params.length - 1];
```


Hopefully gets resolved soon.

**Comment by carbogninalberto** [comment]
I have looked into the support for svelte 5 it will take a bit longer because there are some parts of the code that need a refactor.

I will open the PR soon.

**Comment by mateothegreat** [comment]
I was a full invested svelte-spa-router user until 5 dropped and didn't have a choice but to do my part and throw up https://github.com/mateothegreat/svelte5-router for the community. Hope it unblocks somebody.

Thank you for all your hard work in keeping this lib alive <3

**Comment by brandonkal** [comment]
@mateothegreat that is not appear to be a hash-based router so it is not comparable

**Comment by dvcol** [comment]
@brandonkal If you really need a svelte 5 native hash router, I wrote [something a few month ](https://github.com/dvcol/svelte-simple-router) back.

It's jsut a side project however, so I would be wary of using it in prod.
For serious project, I'd wait for @doregex's PR

**Comment by arturovergara** [comment]
Any update on this? :(

**Comment by Turbocat6413** [comment]
soo, any updates? =)

**Comment by webJose** [comment]
I too am in need of a client-sided router.  I'm working on one since I'm running out of time, and figured I should contribute an option.

I'll put the README here.  I don't have a repo link ready for you guys just yet.  I'll update this post once I push code.  A couple of things:

1. If you would like to provide feedback, please do so.
2. This will **not** include hash routing.  It is my understanding that hash routing used to be the norm when we could not do routing with the URL's pathname.  Correct me if I'm wrong.
3. This is still WIP.  It will take a month or so to be published.

---

# package-name

> Svelte v5 unintrusive, micro client router.

## Features

This router's code is very small, revolving the 500 lines of code, and this includes typing.

### `<Router>` Component

+ **Multi-matching routes**:  All routes are evaluated.
+ **Non-root serving**:  Specify base paths that are inherited by nesting routers.
+ **Nesting routers**:  Add child routers for fine-grained control.
+ **Fallback content**:  Show users something when no routes match.
+ **Liberty**:  Place anything anywhere inside.  No child restrictions.

### `<Route>` Component

+ **Exact path matching**:  Exact match by default; use rest params to relax the condition.
+ **Path as string or regular expression**:  Define paths however's best for you.
+ **Route parameteres**:  Define route parameters inside string paths or regular expression paths.
+ **Rest parameter**:  Collect "the rest" of the path.
+ **Optional parameters**:  Parameters may be specified as optional.
+ **Additional matching logic**:  Add a predicate function to further restrict a route's matching ability.
+ **Path is optional**:  Forgo path specification entirely and handle route matching entirely with code.
+ **Superb Intellisense**:  The route parameters are strongly typed when defining them inside a string path.

### `<Link>` Component

+ **Drop-in replacement**:  Exchange `<a>` tags with `<Link>` tags and you're done.
+ **Specify state**:  Set history state upon hyperlink click.
+ **Active state based on route key**:  Automatically set active state and `aria-current` by specifying the route's key.
+ **Replace or push**:  Select the method for pushing state.

### `location` Global Object

+ **Reactive URL**:  URL object that's always in sync with the browser's URL.
+ **Reactive state**:  Reactive state property that's always in sync with the history state.
+ **Programatic navigation**:  Use the the `navigate()` method to trigger navigation programatically.

#### In Full Mode...

+ **Cancellable `beforeNavigate` event**:  Get notified of navigation events, and cancel when appropriate.
+ **`navigationCancelled` event**:  Get notified whenever navigation is cancelled.
+ **`History` API interception**:  Gain control over the history object to avoid external code from de-synchronizing state.
+ **Micro-frontends**:  Full mode's features are great for micro-frontend scenarios where other routers (from potentially other technologies) might interfere with the router's functionality.

## Quickstart

1. Install the package.
2. Initialize the library.
3. Define the routes inside routers.
4. Modify/Add your navigation links.

### Install the package

```bash
npm i package-name
```

### Initialize the Library

```typescript
// In your main.ts, or somewhere BEFORE any routers are created:

// Regular mode:
import { init } from "package-name";

init();

// Full mode:
import { initFull } from "package-name";

initFull();
```

### Define the Routes

`<Route>`s are added inside `<Router>`s.  `<Router>`s can be nested inside other `<Router>`s.  `<Route>`s can render `<Router>`s or other `<Route>`s, etc.  You get the idea:  You do as you wish.

```svelte
<script lang="ts">
    import { Router, Route } from "package-name";
    import NavBar from "./lib/NavBar.svelte";
    import UserView from "./lib/UserView.svelte";
</script>

<Route>
    <NavBar />
    <div class="container">
        <Router>
            <!-- content outside routes is always rendered -->
            <h1>Route Sample</h1>
            <Route key="users" path="/users">
                <!-- content here -->
            </Route>
            <Route key="user" path="/users/:userId">
                <!-- access parameters via the snippet parameter -->
                {#snippet children(params)}
                    <UserView id={params.userId} /> <!-- Intellisense will work here!! -->
                {/snippet}
            </Route>
            ...
        </Router>
    </div>
</Route>
```

### Navigation Links

The previous step adds the `<NavBar />` component inside the router.  This is the best practice for full `<Link>` functionality.  Still, this is not mandatory.

```svelte
<script lang="ts">
    // NavBar.svelte
    import { Link } from "package-name";
</script>

<nav>
    <div class="nav-links">
        <ul>
            <li class="nav-link">
                <Link href="/users" activeRouteKey="users" activeState={{ class: 'active' }}>All Users</Link>
            </li>
            ...
        </ul>
    </div>
</nav>
```

## Unintrusive Philosophy

This micro router library imposes minimal restrictions.  Here are some features provided by other much larger codebases that are not provided here because Svelte already has the capability, and this package has made sure these aren't blocked.

### Transitions

Nothing prevents you to add transitions to anything.

```svelte
<Route key="users" path="/users/:userId">
    {#snippet children(params)}
        <div transition:fade>
            ...
        </div>
    {/snippet}
</Route>
```

### Guarded Routes

Guard routes however you wish.  Maybe with an `{#if}` block, or maybe using the route's `and` property that allows you to specify a predicate function.  There are probably many other ways.

### Exact Property on Routes

Not needed.  If you want to opt out of the exact route matching, simply add the rest parameter specifier:

```svelte
<Route key="admin" path="/admin/*">
</Route>
```

If you don't care about the value of the parameter, just ignore it.

### Fallback Route

Fallback content is simply specified inside the `fallback` snippet of routers.  There is no need for a fallback `<Route>` component.

### Lazy-Loading

Lazy-loading components is very simple:

```svelte
<script lang="ts">
    function loadUsersComponent() {
        return import('./lib/Users.svelte').then(m => m.default);
    }
</script>

<Route key="users" path="/users">
    {#await loadUsersComponent()}
        <span>Loading...</span>
    {:then Users}
        <Users />
    {:catch}
        <p>Ooops!</p>
    {/await}
</Route>
```

## Reacting To Changes

This is part of the unintrusive philosophy/unneeded features discussion, but because of its importance, it got its own section.

### Reacting to Changes

Other packages offer events.  This is not needed.  React to Svelte state instead:

```typescript
import { location } from "package-name";

// Or $derived, whichever you need.
$effect(() => {
    // Read location.url to re-run on URL changes (navigation).
    location.url;
    // Read location.state to re-run on state changes.
    location.state;
});
```

2Do: Have `<Router>` expose its router engine.

### Reacting to Route Matching Changes

**Comment by brandonkal** [comment]
@webJose Hash routing is still very much in use today (see gmail etc) and is not replaced by new browser APIs. While it is possible to fake a navigation without JavaScript, the server still needs to support that URL which is not ideal. With hash-based routing, the hash part is not sent to the server so the same base html and js will be downloaded without having to configure nginx to treat multiple "virtual urls" as the same. Also replacing the anchor tag with a custom Link component is not ideal because it ignores anchor tags served from CMS systems and those that are part of other existing components.

**Comment by webJose** [comment]
@brandonkal Thanks for your input!  So I stand corrected.  Thank you very much.

Follow-up question:  Is hash routing done uniformly by everyone?  Like `#/regular/path/name/as/it/were/the/actual/path/name`?  Or are various formats/liberty of formats very much on the table?

Then the `<Link>` vs `<a>` sentence:  How do those scenarios work, or people like you would expect them to work?  If the link's `href` changes the URL, the page will reload, correct?  That would be the normal behavior.  If I had a Svelte SPA application using my Router, reloading simply resets all JS, router included.  So what's the expectation from the router library here?  To intercept all anchor tags?

**Comment by db6edr** [comment]
@webJose You can use `<a href="#/new/path">Text</a>`. This doesn't trigger a page reload, it just fires some navigation events. And if there's an element with `id="/new/path"`, it will be scrolled into view.

**Comment by dvcol** [comment]
I'd be weary or using hashchange and popstate events in a micro-frontend context. 

Eventlistener implementation vary between frameworks (angular, vue or react routers) and capturing/emitting such events might cause unwanted side effects.

I had issues in the past where vue router and angular router would trigger infinite loops when navigating from the top shell.

If you want to use it to sync you router(s) state you might need to do some monkeypatching unfortunately.

Alternatively you could use the experimental [navigation API](https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API#browser_compatibility), but without polyfill the support is mostly limited to chromium.

I did use it with some succes in [svelte-simple-router](https://github.com/dvcol/svelte-simple-router/blob/9eb2d4e036aa596312cebf2dac7182f14ab536da/src/lib/router/router.svelte.ts#L381), so your mileage may vary 🤷

**Comment by webJose** [comment]
My event listeners are unintrusive, simply being used to update Svelte state in my router library.  My router implementation, at least currently, does not enter a war trying to update the URL based on the events happening, so I foresee no issues in this particular area.

The full mode that intercepts the History API is just done to update state whenever child routers push or replace state, which is an operation that doesn't trigger any events.  I think I'm covered on these fronts.  Of course, this is all theoretical so far.  I haven't gotten to the experimentation part.

The one front I have a potential problem is when my router pushes state.  This doesn't produce any events, and I need to investigate what routers do.  However, I suspect that I need to synthesize a `popstate` event so child routers update themselves.  Still, I think that this should not be needed because I think micro-frontends should not be meddling with sub-paths of URL's that belong to other micro-frontends.  Example: MFE A mounts in the root path "/".  Its router would mount MFE B whenever the path matches "/mfeB/*".  There should be **zero** links provided by MFE A whose `href` property has a URL with path segments beyond the `/mfeB` segment.  So I think this potential problem only arises when good practices aren't followed.

**Comment by webJose** [comment]
## Update

Ok, so I have much of the router ready.  It has unit testing, and all that remains would be to add support for hash routing.  There is, however, an issue I would like to address, if possible:  Route parameters.

I have route parameters implemented, and they are specified in the Express format that all JavaScripters seem to love above all:  `:<parameter name>`.  Furthermore, I have added TypeScript types that can infer the parameter names solely from the path:

![Typed parameters](https://github.com/user-attachments/assets/6655a4c5-5af5-4b31-90cf-f89e7b28bec5)

As seen in the screenshot, by defining the parameter inside the `path` prop, VS Code will provide Intellisense on the declared `step` parameter.  Pretty need.  However, there are things that TS cannot do that the parameter mechanism can.  Example:  `/quickstart/:step?-detailed` would work in JS, but TS cannot match this ability, and the inferred parameter name would not be `step`, it would be `step?-detailed`.

So I am requesting feedback:  I want to move to the .Net format `{<parameter name>}` because the use a begining and ending character will suit TS and Intellisense better.

Let me know your thoughts as potential users of the library.  Thanks!

**Comment by webJose** [comment]
## Final Update Before Release

Ok, this thread has gone silent.  Unsure if there's still interest in a new routing solution for Svelte v5.  Let's see if I can change that. 😄 

I'm pretty much done with coding and basic testing.  I think I can release an experimental version now for all interested parties to test-drive.

### Presenting:  Simultaneous and Multiple Routing

My router can simultaneously route using the URL's path name (called path routing) and the URL's hash value (called hash routing).  But not just that:  When doing hash routing, the hash itself can contain multiple paths!  You can theoretically have unlimited routes simultaneously:  One for path routing, and the rest in hash.

You can potentially mount side-by-side micro-frontends that route independently and simultaneously.  The demo in the project (that I will upload to GitHub) does this:

![Image](https://github.com/user-attachments/assets/ef8e5ea6-ac21-442f-b1b8-198fea666651)

**NOTE**:  4K screenshot there.  Apologies if the image is too large.

What you see is the same component (`DemoView` is its name, I think) loaded twice.  One uses the `d1` path and the other the `d2` path.  And where are those paths defined?  In the hash value.  The URL for this demo is http://localhost:5174/hash-routing#d1=/demo;d2=/demo/location.

### Lines of Code

As of this writing:

```plaintext
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Svelte                           5             53              5            578
TypeScript                      10             49            335            461
Markdown                         2             56              0            246
-------------------------------------------------------------------------------
SUM:                            17            158            340           1285
-------------------------------------------------------------------------------
```

Excluding Markdown, 1039 LOC.  Not bad, but I was hoping to ship less than 1000.  It is what it is.

**Comment by webJose** [comment]
## @wjfe/n-savant

Hello!  I just published the first version: v0.1.0.  Remember:  This is **experimental** and I have only done limited testing.  Yes, it seems to work nicely, but I cannot know this if you guys don't report encountered issues.  Open source software is great so long **we participate**.  Please participate providing feedback.

[Online Documentation](https://wjfe-n-savant.hashnode.space/)
[Repository](https://github.com/WJSoftware/wjfe-n-savant)
[Issues](https://github.com/WJSoftware/wjfe-n-savant/issues)

I also enabled [Discussions](https://github.com/WJSoftware/wjfe-n-savant/discussions).  Feel free to ask questions, how-to's, share your experience, anything.

**Comment by carbogninalberto** [comment]
Sorry guys for the significant delay I was too busy. Here the PR #331

**Comment by ivobar** [comment]
> Sorry guys for the significant delay I was too busy. Here the PR [#331](https://github.com/ItalyPaleAle/svelte-spa-router/pull/331)

Heya, is there any prospect of moving along with the PR ? Do you need help ? We've been using this router and want to migrate to Svelte 5 but there are some problems with reactivity even in compatibility mode and I would hate to do monkey patching :-)

**Comment by carbogninalberto** [comment]
Hi @ivobar as soon as I have time I will finish the PR, it's almost working, there are some corner issues I need to figure it out. What kind of problem do you have in the reactivity?

This would help to identify potential issues in the work I am doing, thanks!

**Comment by ivobar** [comment]
> Hi [@ivobar](https://github.com/ivobar) as soon as I have time I will finish the PR, it's almost working, there are some corner issues I need to figure it out. What kind of problem do you have in the reactivity?
> 
> This would help to identify potential issues in the work I am doing, thanks!

@carbogninalberto Heya,

We are currently investigating what the problem is but I've managed to narrow it down to an issue that is not related to the router, sorry for the confusion. For now the router works as expected with the compiler compatibility mode set to Svelte 4.

**Comment by pjaudiomv** [comment]
No, the answer is no. This library will never support Svelte 5.

**Comment by carbogninalberto** [comment]
Yes done it https://github.com/ItalyPaleAle/svelte-spa-router/pull/331 I am waiting for the review of Alessandro to merge it. Sorry guys for the long time it took me to finish it, I have been pretty busy lately.
